File: apps/api/app/schemas/invoice.py
from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from app.db.models.invoice import InvoiceStatus
from app.schemas.common import BaseReadModel
from app.schemas.quote import BillingLineItemInput, BillingLineItemRead


BillingFollowUpStatus = Literal["normal", "to_follow_up", "followed_up", "waiting_customer"]


class InvoiceRead(BaseReadModel):
    organization_id: UUID
    customer_id: UUID
    customer_name: str
    worksite_id: UUID | None
    worksite_name: str | None
    sequence_number: int
    number: str
    title: str | None
    issue_date: date
    due_date: date | None
    status: InvoiceStatus
    follow_up_status: BillingFollowUpStatus
    currency: str
    line_items: list[BillingLineItemRead]
    subtotal_amount_cents: int
    total_amount_cents: int
    paid_amount_cents: int
    paid_at: date | None
    outstanding_amount_cents: int
    notes: str | None


class InvoiceCreateRequest(BaseModel):
    customer_id: UUID
    worksite_id: UUID | None = None
    title: str | None = Field(default=None, max_length=160)
    issue_date: date
    due_date: date | None = None
    status: InvoiceStatus = InvoiceStatus.DRAFT
    currency: str | None = Field(default="EUR", min_length=3, max_length=3)
    line_items: list[BillingLineItemInput] = Field(min_length=1)
    notes: str | None = Field(default=None, max_length=2000)

    @model_validator(mode="after")
    def validate_dates(self) -> "InvoiceCreateRequest":
        if self.due_date is not None and self.due_date < self.issue_date:
            raise ValueError("La date d'echeance doit etre posterieure ou egale a la date d'emission.")
        return self


class InvoiceUpdateRequest(BaseModel):
    customer_id: UUID
    worksite_id: UUID | None = None
    title: str | None = Field(default=None, max_length=160)
    issue_date: date
    due_date: date | None = None
    line_items: list[BillingLineItemInput] = Field(min_length=1)
    notes: str | None = Field(default=None, max_length=2000)

    @model_validator(mode="after")
    def validate_dates(self) -> "InvoiceUpdateRequest":
        if self.due_date is not None and self.due_date < self.issue_date:
            raise ValueError("La date d'echeance doit etre posterieure ou egale a la date d'emission.")
        return self


class InvoiceStatusUpdateRequest(BaseModel):
    status: InvoiceStatus


class InvoicePaymentCreateRequest(BaseModel):
    paid_amount_cents: int = Field(gt=0)
    paid_at: date


class InvoiceWorksiteLinkUpdateRequest(BaseModel):
    worksite_id: UUID | None = None


class InvoiceFollowUpUpdateRequest(BaseModel):
    follow_up_status: BillingFollowUpStatus

File: packages/contracts/src/invoice.ts
import type { EntityId, VersionedRecord } from "./common";
import type { BillingFollowUpStatus, BillingFollowUpUpdateRequest } from "./billing-follow-up";
import type { BillingLineItemInput, BillingLineItemRecord } from "./billing-line";

export type InvoiceStatus = "draft" | "issued" | "paid" | "overdue";

export interface InvoiceRecord extends VersionedRecord {
  organization_id: EntityId;
  customer_id: EntityId;
  customer_name: string;
  worksite_id: EntityId | null;
  worksite_name: string | null;
  sequence_number: number;
  number: string;
  title: string | null;
  issue_date: string;
  due_date: string | null;
  status: InvoiceStatus;
  follow_up_status: BillingFollowUpStatus;
  currency: string;
  line_items: BillingLineItemRecord[];
  subtotal_amount_cents: number;
  total_amount_cents: number;
  paid_amount_cents: number;
  paid_at: string | null;
  outstanding_amount_cents: number;
  notes: string | null;
}

export interface InvoiceCreateRequest {
  customer_id: EntityId;
  worksite_id?: EntityId | null;
  title?: string | null;
  issue_date: string;
  due_date?: string | null;
  status?: InvoiceStatus;
  currency?: string | null;
  line_items: BillingLineItemInput[];
  notes?: string | null;
}

export interface InvoiceUpdateRequest {
  customer_id: EntityId;
  worksite_id?: EntityId | null;
  title?: string | null;
  issue_date: string;
  due_date?: string | null;
  line_items: BillingLineItemInput[];
  notes?: string | null;
}

export interface InvoiceStatusUpdateRequest {
  status: Extract<InvoiceStatus, "draft" | "issued">;
}

export interface InvoicePaymentCreateRequest {
  paid_amount_cents: number;
  paid_at: string;
}

export interface InvoiceWorksiteLinkUpdateRequest {
  worksite_id?: EntityId | null;
}

export interface InvoiceFollowUpUpdateRequest extends BillingFollowUpUpdateRequest {}
